File: sync.js
const simpleGit = require('simple-git');
const { Octokit } = require('@octokit/rest');
const fs = require('fs');
const path = require('path');

const CONFIG_FILE = path.join(__dirname, 'sync-config.json');
const TEMP_DIR = path.join(__dirname, 'temp');

const loadConfig = () => {
  if (!fs.existsSync(CONFIG_FILE)) {
    console.error('❌ 配置文件不存在: ' + CONFIG_FILE);
    process.exit(1);
  }

  try {
    const content = fs.readFileSync(CONFIG_FILE, 'utf8');
    const config = JSON.parse(content);
    
    if (!config.sourceRepos || !Array.isArray(config.sourceRepos) || config.sourceRepos.length === 0) {
      console.error('❌ 配置文件中 sourceRepos 数组为空或格式错误');
      process.exit(1);
    }
    
    return config;
  } catch (error) {
    console.error('❌ 读取配置文件失败: ' + error.message);
    process.exit(1);
  }
};

const getRepoInfo = (url) => {
  const match = url.match(/https:\/\/github\.com\/(.*?)\/(.*?)(?:\.git)?$/);
  if (!match) {
    throw new Error('无效的 GitHub 仓库 URL: ' + url);
  }
  return { owner: match[1], repo: match[2] };
};

const getDefaultBranch = async (octokit, owner, repo) => {
  try {
    const result = await octokit.repos.get({ owner, repo });
    return result.data.default_branch;
  } catch (error) {
    console.error('   ❌ 获取默认分支失败: ' + error.message);
    throw error;
  }
};

const cloneRepository = async (url, targetDir, branch) => {
  const git = simpleGit();
  
  try {
    await git.clone(url, targetDir, {
      '--single-branch': true,
      '--branch': branch
    });
    return branch;
  } catch (error) {
    if (error.message.includes('not found') || error.message.includes('does not exist')) {
      console.log(`   ⚠️  分支 ${branch} 不存在，尝试获取默认分支`);
      return null;
    }
    throw error;
  }
};

const removeWorkflows = async (repoGit) => {
  const workflowsDir = path.join(repoGit.cwd, '.github', 'workflows');
  if (fs.existsSync(workflowsDir)) {
    console.log('   🗑️  删除 .github/workflows 目录（避免权限问题）');
    fs.rmSync(workflowsDir, { recursive: true, force: true });
    
    try {
      await repoGit.rm('-r', '--cached', '.github/workflows');
      console.log('   🧹 从 Git 索引中移除工作流文件');
    } catch (error) {
      console.log('   ⚠️  清理 Git 索引时出错（可忽略）:', error.message);
    }
  }
};

const syncRepository = async (repoConfig, octokit, targetRepo, githubToken) => {
  const { repoUrl, sourceBranch, targetBranch, repoName } = repoConfig;
  const tempDir = path.join(TEMP_DIR, repoName);
  
  console.log(`\n🔄 同步仓库: ${repoName}`);
  console.log(`   源: ${repoUrl} (${sourceBranch})`);
  console.log(`   目标: ${targetRepo} (${targetBranch})`);

  let actualBranch = sourceBranch;

  try {
    actualBranch = await cloneRepository(repoUrl, tempDir, actualBranch);
    
    if (!actualBranch) {
      const { owner, repo } = getRepoInfo(repoUrl);
      actualBranch = await getDefaultBranch(octokit, owner, repo);
      console.log(`   🔍 找到默认分支: ${actualBranch}`);
      
      await cloneRepository(repoUrl, tempDir, actualBranch);
    }

    const repoGit = simpleGit(tempDir);
    await repoGit.pull('origin', actualBranch);
    
    await removeWorkflows(repoGit);

    const targetUrl = `https://x-access-token:${githubToken}@github.com/${targetRepo}.git`;
    await repoGit.push(targetUrl, `${actualBranch}:${targetBranch}`, {
      '--force': true
    });

    console.log(`   ✅ 同步成功: ${actualBranch} -> ${targetBranch}`);

  } catch (error) {
    console.error(`   ❌ 同步失败: ${error.message}`);
    throw error;
  } finally {
    if (fs.existsSync(tempDir)) {
      fs.rmSync(tempDir, { recursive: true, force: true });
    }
  }
};

const main = async () => {
  const config = loadConfig();
  const { sourceRepos, targetRepo } = config;
  const githubToken = process.env.GITHUB_TOKEN;

  if (!githubToken) {
    console.error('❌ 缺少 GITHUB_TOKEN 环境变量');
    process.exit(1);
  }

  if (!targetRepo) {
    console.error('❌ 配置文件中缺少 targetRepo 字段');
    process.exit(1);
  }

  const octokit = new Octokit({ auth: githubToken });

  console.log('🚀 开始同步仓库...');
  console.log(`📋 目标仓库: ${targetRepo}`);
  console.log(`📦 源仓库数量: ${sourceRepos.length}`);
  console.log('💡 提示: 已自动排除 .github/workflows 目录以避免权限问题');

  let successCount = 0;
  let failCount = 0;

  for (const repoConfig of sourceRepos) {
    try {
      await syncRepository(repoConfig, octokit, targetRepo, githubToken);
      successCount++;
    } catch (error) {
      failCount++;
      console.error(`\n❌ 仓库 ${repoConfig.repoName} 同步失败，继续处理下一个仓库`);
    }
  }

  console.log('\n📊 同步统计:');
  console.log(`   ✅ 成功: ${successCount}`);
  console.log(`   ❌ 失败: ${failCount}`);
  console.log(`   📦 总计: ${sourceRepos.length}`);

  if (failCount > 0) {
    console.log('\n⚠️  部分仓库同步失败');
    process.exit(1);
  } else {
    console.log('\n🎉 所有仓库同步成功！');
  }
};

main().catch(error => {
  console.error('\n❌ 程序异常:', error.message);
  console.error(error.stack);
  process.exit(1);
});

File: sync.py
#!/usr/bin/env python3
import os
import json
import shutil
import subprocess
import sys
from pathlib import Path
from github import Github

CONFIG_FILE = Path(__file__).parent / 'sync-config.json'
TEMP_DIR = Path(__file__).parent / 'temp'

def load_config():
    """加载配置文件"""
    if not CONFIG_FILE.exists():
        print(f'❌ 配置文件不存在: {CONFIG_FILE}')
        sys.exit(1)
    
    try:
        with open(CONFIG_FILE, 'r', encoding='utf-8') as f:
            config = json.load(f)
        
        if not config.get('sourceRepos') or not isinstance(config['sourceRepos'], list) or len(config['sourceRepos']) == 0:
            print('❌ 配置文件中 sourceRepos 数组为空或格式错误')
            sys.exit(1)
        
        return config
    except Exception as e:
        print(f'❌ 读取配置文件失败: {e}')
        sys.exit(1)

def get_repo_info(url):
    """解析仓库URL"""
    import re
    match = re.match(r'https://github\.com/(.*?)/(.*?)(?:\.git)?$', url)
    if not match:
        raise ValueError(f'无效的 GitHub 仓库 URL: {url}')
    return {'owner': match.group(1), 'repo': match.group(2)}

def get_default_branch(github_client, owner, repo):
    """获取仓库的默认分支"""
    try:
        repo_obj = github_client.get_repo(f'{owner}/{repo}')
        return repo_obj.default_branch
    except Exception as e:
        print(f'   ❌ 获取默认分支失败: {e}')
        raise

def clone_repository(url, target_dir, branch):
    """克隆仓库"""
    try:
        subprocess.run([
            'git', 'clone', '--single-branch', '--branch', branch, url, str(target_dir)
        ], check=True, capture_output=True, text=True)
        return branch
    except subprocess.CalledProcessError as e:
        if 'not found' in e.stderr or 'does not exist' in e.stderr:
            print(f'   ⚠️  分支 {branch} 不存在，尝试获取默认分支')
            return None
        raise

def remove_workflows(repo_dir):
    """删除工作流文件"""
    workflows_dir = repo_dir / '.github' / 'workflows'
    if workflows_dir.exists():
        print('   🗑️  删除 .github/workflows 目录（避免权限问题）')
        
        try:
            result = subprocess.run([
                'git', 'ls-files', '.github/workflows'
            ], check=False, capture_output=True, text=True, cwd=str(repo_dir))
            
            if result.stdout.strip():
                subprocess.run([
                    'git', 'rm', '-r', '-f', '.github/workflows'
                ], check=True, capture_output=True, text=True, cwd=str(repo_dir))
                print('   🧹 从 Git 索引中移除工作流文件')
            else:
                print('   ℹ️  .github/workflows 目录不在 Git 索引中，直接删除')
        except subprocess.CalledProcessError as e:
            print(f'   ⚠️  清理 Git 索引时出错（可忽略）: {e}')
        
        shutil.rmtree(workflows_dir)

def sync_repository(repo_config, github_client, target_repo, github_token):
    """同步单个仓库"""
    repo_url = repo_config['repoUrl']
    source_branch = repo_config.get('sourceBranch', 'main')
    target_branch = repo_config.get('targetBranch', repo_config.get('repoName', ''))
    repo_name = repo_config['repoName']
    
    temp_dir = TEMP_DIR / repo_name
    
    print(f'\n🔄 同步仓库: {repo_name}')
    print(f'   源: {repo_url} ({source_branch})')
    print(f'   目标: {target_repo} ({target_branch})')
    
    actual_branch = source_branch
    
    try:
        actual_branch = clone_repository(repo_url, temp_dir, actual_branch)
        
        if actual_branch is None:
            repo_info = get_repo_info(repo_url)
            actual_branch = get_default_branch(github_client, repo_info['owner'], repo_info['repo'])
            print(f'   🔍 找到默认分支: {actual_branch}')
            
            clone_repository(repo_url, temp_dir, actual_branch)
        
        subprocess.run([
            'git', 'pull', 'origin', actual_branch
        ], check=True, capture_output=True, text=True, cwd=str(temp_dir))
        
        remove_workflows(temp_dir)
        
        target_url = f'https://x-access-token:{github_token}@github.com/{target_repo}.git'
        subprocess.run([
            'git', 'push', '--force', target_url, f'{actual_branch}:{target_branch}'
        ], check=True, capture_output=True, text=True, cwd=str(temp_dir))
        
        print(f'   ✅ 同步成功: {actual_branch} -> {target_branch}')
        
    except Exception as e:
        print(f'   ❌ 同步失败: {e}')
        raise
    finally:
        if temp_dir.exists():
            shutil.rmtree(temp_dir)

def main():
    """主函数"""
    config = load_config()
    source_repos = config['sourceRepos']
    target_repo = config['targetRepo']
    github_token = os.getenv('GITHUB_TOKEN')
    
    if not github_token:
        print('❌ 缺少 GITHUB_TOKEN 环境变量')
        sys.exit(1)
    
    if not target_repo:
        print('❌ 配置文件中缺少 targetRepo 字段')
        sys.exit(1)
    
    github_client = Github(github_token)
    
    print('🚀 开始同步仓库...')
    print(f'📋 目标仓库: {target_repo}')
    print(f'📦 源仓库数量: {len(source_repos)}')
    print('💡 提示: 已自动排除 .github/workflows 目录以避免权限问题')
    
    success_count = 0
    fail_count = 0
    
    for repo_config in source_repos:
        try:
            sync_repository(repo_config, github_client, target_repo, github_token)
            success_count += 1
        except Exception as e:
            fail_count += 1
            print(f'\n❌ 仓库 {repo_config["repoName"]} 同步失败，继续处理下一个仓库')
    
    print('\n📊 同步统计:')
    print(f'   ✅ 成功: {success_count}')
    print(f'   ❌ 失败: {fail_count}')
    print(f'   📦 总计: {len(source_repos)}')
    
    if fail_count > 0:
        print('\n⚠️  部分仓库同步失败')
        sys.exit(1)
    else:
        print('\n🎉 所有仓库同步成功！')

if __name__ == '__main__':
    main()